LINUX/if_vmxnet3_netmap.h

```diff
@@ -0,0 +1,593 @@
+
+	union Vmxnet3_GenericDesc *gdesc;
+	struct SOFTC_T *adapter = netdev_priv(ifp);
+	struct vmxnet3_tx_queue *tq = &adapter->tx_queue[ring_nr];
+
+	if (!netif_carrier_ok(ifp))
+		return 0;
+
+	//
+	// Free up the comp_descriptors aggressively
+	//
+
+	completed = vmxnet3_netmap_tq_tx_complete(tq, adapter->pdev);
+
+	//
+	// Reclaim buffers for completed transmissions
+	//
+
+	kring->nr_hwtail =
+		nm_prev(tq->comp_ring.next2proc, tq->comp_ring.size - 1);
+
+	//
+	// Process new packets to send
+	//
+
+	nm_i = kring->nr_hwcur;
+
+	if (nm_i != head) {
+		for (n = 0; nm_i != head; n++) {
+			u32 copy_size = 0;
+			int free_cmd_desc_count;
+			unsigned long lock_flags;
+
+			union Vmxnet3_GenericDesc *sop_txd;
+			union Vmxnet3_GenericDesc *eop_txd;
+
+			struct netmap_slot *slot = &ring->slot[nm_i];
+			u_int packet_len = slot->len;
+			void *packet_addr = NMB(na, slot);
+
+			slot->flags &= ~(NS_REPORT | NS_BUF_CHANGED);
+
+			spin_lock_irqsave(&tq->tx_lock, lock_flags);
+
+			free_cmd_desc_count =
+				vmxnet3_cmd_ring_desc_avail(&tq->tx_ring);
+
+			if (free_cmd_desc_count < kMinFreeTxDescForPacket) {
+				tq->stats.tx_ring_full++;
+				spin_unlock_irqrestore(&tq->tx_lock,
+						       lock_flags);
+				break;
+			}
+
+			//
+			// Copy header
+			//
+
+			if (kUseTwoTxDescForPacket) {
+				struct Vmxnet3_TxDataDesc *tdd = tdd =
+					tq->data_ring.base +
+					tq->tx_ring.next2fill;
+
+				copy_size = min((u_int)VMXNET3_HDR_COPY_SIZE,
+						packet_len);
+				memcpy(tdd->data, packet_addr, copy_size);
+			}
+
+			//
+			// Map rest of data
+			//
+
+			{
+				u32 dw2;
+				u32 len;
+				u32 buf_offset;
+				union Vmxnet3_GenericDesc *gdesc;
+				struct vmxnet3_tx_buf_info *tbi = NULL;
+
+				//
+				// use the previous gen bit for the SOP desc
+				//
+
+				dw2 = (tq->tx_ring.gen ^ 0x1)
+				      << VMXNET3_TXD_GEN_SHIFT;
+
+				sop_txd = tq->tx_ring.base +
+					  tq->tx_ring.next2fill;
+				gdesc = sop_txd;
+
+				//
+				// Setup TX descriptor for the header
+				//
+
+				if (copy_size) {
+					sop_txd->txd.addr = cpu_to_le64(
+						tq->data_ring.basePA +
+						tq->tx_ring.next2fill *
+							sizeof(struct
+							       Vmxnet3_TxDataDesc));
+					sop_txd->dword[2] =
+						cpu_to_le32(dw2 | copy_size);
+					sop_txd->dword[3] = 0;
+
+					tbi = tq->buf_info +
+					      tq->tx_ring.next2fill;
+					tbi->map_type = VMXNET3_MAP_NONE;
+
+					vmxnet3_cmd_ring_adv_next2fill(
+						&tq->tx_ring);
+
+					//
+					// use the right gen for non-SOP desc
+					//
+
+					dw2 = tq->tx_ring.gen
+					      << VMXNET3_TXD_GEN_SHIFT;
+				}
+
+				//
+				// Handle linear part
+				//
+
+				len = packet_len - copy_size;
+				buf_offset = copy_size;
+
+				if (len) {
+					u32 buf_size;
+
+					BUG_ON(len > VMXNET3_MAX_TX_BUF_SIZE);
+					buf_size = len;
+					dw2 |= len;
+
+					tbi = tq->buf_info +
+					      tq->tx_ring.next2fill;
+					tbi->map_type = VMXNET3_MAP_SINGLE;
+					tbi->dma_addr = dma_map_single(
```

Comment:
All netmap buffers are already DMA-mapped bidirectionally, so you should not call dma_map_single again. ```PNMB``` (see above) will give you the ```dma_addr``` you are looking for. You also need to call ```netmap_sync_map``` to flush caches as needed. Please take a look at ````ixgbe_netmap_linux.h``` for examples.